cloudinit/stages.py

```diff
@@ -63,6 +75,132 @@
 )
 
 
+@dataclass
+class Semaphore:
+    semaphore: helpers.FileSemaphores
+    args: Tuple[str, str]
+
+
+class CacheMode(str, Enum):
+    trust = "trust"
+    check = "check"
+
+
+# used in status.json and some log strings
+UserLongName = Literal[
+    "Local Stage",
+    "Network Stage",
+    "Config Stage",
+    "Final Stage",
+]
+
+# used in status.json and some log strings
+UserShortName = Literal[
+    "init-local",
+    "init",
+    "modules-config",
+    "modules-final",
+]
+
+# The first argument in main_init() / main_modules() / main_single()
+ArgName = Literal[
+    "init",
+    "modules",
+    "single",
+]
+
+# The first argument in main_init() / main_modules() / main_single()
+InternalName = Literal[
+    "local",
+    "network",
+    "config",
+    "final",
+]
+InternalNameSingle = Union[InternalName, Literal["single"]]
+
+# Because code reuse is evil and user confusion is the goal.
+#
+# Kidding. Lets please standardize but I've gotten pushback on changing this
+# string in the past so for now I guess we get yet another user-facing string
+# that means the same thing.
+WelcomeName = Literal[
+    "init-local", "init", "modules:config", "modules:final", "single"
+]
+
+
+@dataclass
+class BaseStage:
+    welcome: WelcomeName
+    name: ArgName
+    deps: List[sources.Deps]
+
+
+@dataclass
+class BootStage(BaseStage):
+    """BootStage is used by static analyzers to ensure that values are valid
+
+    This type excludes "single" which enables stricter type checking. Where
+    "single" is also required, use "Stage" which is a union of SingleStage and
+    BootStage.
+
+    long: user-facing stage name
+    short: status.json stage name, used in some logs
+    """
+
+    short: UserShortName
+    long: UserLongName
+    internal: InternalName
+
+
+@dataclass
+class SingleStage(BaseStage):
+    internal: InternalNameSingle
+
+
+Stage = Union[SingleStage, BootStage]
+
+# These module variables store static data that is associated with each stage.
+# https://docs.cloud-init.io/en/latest/explanation/boot.html
+local: Final = BootStage(
```

Comment:
These dataclasses hold all the (formerly) magic strings. Note that dsmode is not included because dsmode is a datasource & runtime concept, not a thing that is known before running the code.